File: Scripts/Python/Sep_03/bernstein.py
#!/usr/bin/env python
# File: bernstein.py
# Name: D.Saravanan
# Date: 03/09/2021

""" Script to compute Bernstein polynomial """

import numpy as np
import numpy.polynomial.polynomial as poly
from scipy.special import legendre
import matplotlib
matplotlib.rcParams['pgf.texsystem'] = 'pdflatex'
matplotlib.rcParams.update({'font.family': 'serif', 'font.size': 8,
    'axes.labelsize': 10, 'axes.titlesize': 10, 'figure.titlesize': 10})
matplotlib.rcParams['text.usetex'] = True
import matplotlib.pyplot as plt
matplotlib.use('Agg')

def f(x):
    return 1/(1 + 25 * x**2)

def factorial(n):
    return 1 if n < 2 else n*factorial(n-1)

M = 1001
x = np.linspace(-1,1,M)
y = f(x)

N = 35

Bern = np.zeros(M)

for k in range(0,N+1):
    Bern = Bern + factorial(N)/(factorial(N-k)*factorial(k)) * ((x+1)/2)**k * ((1-x)/2)**(N-k) * f(2*k/N-1)

fig, ax = plt.subplots()
ax.plot(x, y, c='red', ls='dotted', label=r'$\displaystyle\frac{1}{1+25 x^{2}}$')
ax.plot(x, Bern, c='teal', label=r'$Bernstein$')
ax.set(xlim=(-1, 1), ylim=(0, 1.1))
ax.set(xlabel=r'$x$', ylabel=r'$\displaystyle\frac{1}{1+25 x^{2}}$')
ax.set_title(r'Bernstein polynomial')
ax.grid(True); ax.legend()
plt.savefig('bernstein.pdf')

File: Scripts/Julia/Sep_03/bernstein.jl
#!/usr/bin/env julia
# File: bernstein.jl
# Name: D.Saravanan
# Date: 03/09/2021

""" Script to compute Bernstein polynomial """

import PyPlot
const plt = PyPlot
plt.rc("pgf", texsystem="pdflatex")
plt.rc("font", family="serif", weight="normal", size=8)
plt.rc("axes", labelsize=10, titlesize=10)
plt.rc("figure", titlesize=10)
plt.rc("text", usetex="True")
using LaTeXStrings

function fact(x)
    factorial(big(x))
end

function f(x)
    1 ./(1 .+ 25 .* x.^2)
end

function bernstein(x, M, N)

    bern = zeros(M)

    for k in 0:N
        bern = bern + fact(N)/(fact(N - k)*fact(k)) .* 
                ((x .+ ones(M))./2).^k * ((ones(M) .- x)./2).^(N-k) .* f(2*k/N-1)
    end

    return bern
end

M = 1001
x = LinRange(-1,1,M)
y1 = f(x)

N = 5
y2 = bernstein(x, M, N)
  
fig, ax = plt.subplots()
ax.plot(x, y1, c="olive", ls="dotted", label=L"$\displaystyle\frac{1}{1+25 x^{2}}$")
ax.plot(x, y2, c="teal", label=L"$N = 5$")
ax.grid(true); ax.legend()
plt.savefig("bernstein.png")
